llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -511,12 +604,40 @@ void validateUsedAspectsForFunctions(const FunctionToAspectsMapTy &Map,
   }
 }
 
+struct FunctionsClassifiedByTheirFunctionCall {
+  FunctionSetTy CalledUnconditionally;
+  FunctionSetTy CalledConditionally;
+};
+
+void addAllCalledFunctionsRecursivelyToFunctionsClassifiedByTheirFunctionCallStruct(
+    Function *F, const CallGraphTy &CG,
+    FunctionsClassifiedByTheirFunctionCall &Funcs,
+    SmallPtrSet<const Function *, 16> &Visited) {
+  const auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+  for (Function *Callee : It->second) {
+    if (!Callee->hasFnAttribute("sycl-call-if-on-device-conditionally") &&
+        Funcs.CalledUnconditionally.contains(F)) {
```

Comment:
Results of `.contains` lookup should be the same for all loop iterations if I understand it correctly. Therefore, it should be just called once outside the loop and stored into a `bool` variable. The same applies to the `if` below (lookup into `CalledConditionally`)